app/v2/notifications/post_notifications.py

```diff
@@ -256,32 +259,79 @@ def process_sms_or_email_notification(*, form, notification_type, api_key, templ
+    notification = {
+        "id": create_uuid(),
+        "template": str(template.id),
+        "template_version": str(template.version),
+        "to": form_send_to,
+        "personalisation": personalisation,
+        "simulated": simulated,
+        "api_key_id": str(api_key.id),
+        "key_type": str(api_key.key_type),
+        "client_reference": form.get("reference", None),
+    }
+
+    encrypted_notification_data = encryption.encrypt(notification)
 
     scheduled_for = form.get("scheduled_for", None)
     if scheduled_for:
+        notification = persist_notification(
+            template_id=template.id,
+            template_version=template.version,
+            recipient=form_send_to,
+            service=service,
+            personalisation=personalisation,
+            notification_type=notification_type,
+            api_key_id=api_key.id,
+            key_type=api_key.key_type,
+            client_reference=form.get("reference", None),
+            simulated=simulated,
+            reply_to_text=reply_to_text,
+        )
         persist_scheduled_notification(notification.id, form["scheduled_for"])
     else:
-        if not simulated:
+        if current_app.config["FF_NOTIFICATION_CELERY_PERSISTANCE"]:
+            # depending on the type route to the appropriate save task
+            if notification_type == EMAIL_TYPE:
+                current_app.logger.info("calling save email task")
+                save_email.apply_async(
+                    (authenticated_service.id, create_uuid(), encrypted_notification_data),
+                    queue=QueueNames.DATABASE if not authenticated_service.research_mode else QueueNames.RESEARCH_MODE,
+                )
+            elif notification_type == SMS_TYPE:
+                save_sms.apply_async(
+                    (authenticated_service.id, create_uuid(), encrypted_notification_data),
+                    queue=QueueNames.DATABASE if not authenticated_service.research_mode else QueueNames.RESEARCH_MODE,
+                )
+        else:
+            notification = persist_notification(
+                template_id=template.id,
+                template_version=template.version,
+                recipient=form_send_to,
+                service=service,
+                personalisation=personalisation,
+                notification_type=notification_type,
+                api_key_id=api_key.id,
+                key_type=api_key.key_type,
+                client_reference=form.get("reference", None),
+                simulated=simulated,
+                reply_to_text=reply_to_text,
+            )
             send_notification_to_queue(
                 notification=notification,
                 research_mode=service.research_mode,
                 queue=template.queue_to_use(),
             )
-        else:
-            current_app.logger.debug("POST simulated notification for id: {}".format(notification.id))
+
+    if not isinstance(notification, Notification):
```

Comment:
Potentially could create a function. But I didn't think something like this will be needed anywhere else and its only about 5 lines. Didn't think creating a separate function with only one caller was worth it since it doesn't change readability. But happy to discuss

Reply:
I don't mind much, It's up to you and a soft suggestion. I like recommendations in the [Clean Code](https://www.amazon.ca/dp/0132350882/ref=cm_sw_em_r_mt_dp_S0H2F7ZBC01TBP56YGNV) book, especially to reduce cognitive load and increase code readability. Ideally, if a function could have between 5-10 lines, it would make the code easier to understand. For example and with 10 less lines in the current function:

```
if not isinstance(notification, Notification):
    notification = to_notification_type(notification)
```